src/Adapter/Cart/CommandHandler/UpdateCartDeliverySettingsHandler.php

```diff
@@ -139,4 +123,112 @@ private function createCartRule(Cart $cart, string $backOfficeOrderCode): CartRu
 
         return $freeShippingCartRule;
     }
+
+    /**
+     * @param Cart $cart
+     * @param UpdateCartDeliverySettingsCommand $command
+     *
+     * @throws CannotDeleteCartRuleException
+     * @throws CartRuleException
+     * @throws PrestaShopException
+     */
+    protected function handleFreeShippingOption(Cart $cart, UpdateCartDeliverySettingsCommand $command): void
+    {
+        $backOfficeOrderCode = sprintf('%s%s', CartRule::BO_ORDER_CODE_PREFIX, $cart->id);
+
+        $freeShippingCartRule = $this->getCartRuleForBackOfficeFreeShipping($backOfficeOrderCode);
+
+        if ($command->allowFreeShipping()) {
+            if (null === $freeShippingCartRule) {
+                $freeShippingCartRule = $this->createCartRule($cart, $backOfficeOrderCode);
+            }
+            $cart->addCartRule((int) $freeShippingCartRule->id);
+
+            return;
+        }
+
+        if (null === $freeShippingCartRule) {
+            return;
+        }
+
+        $cart->removeCartRule((int) $freeShippingCartRule->id);
+
+        try {
+            if (false === $freeShippingCartRule->delete()) {
+                throw new CannotDeleteCartRuleException(sprintf('Failed deleting cart rule #%d', $freeShippingCartRule->id));
+            }
+        } catch (PrestaShopException $e) {
+            throw new CartRuleException(sprintf('An error occurred when trying to delete cart rule #%d', $freeShippingCartRule->id));
+        }
+    }
+
+    /**
+     * @param Cart $cart
+     * @param UpdateCartDeliverySettingsCommand $command
+     *
+     * @throws CartException
+     * @throws PrestaShopException
+     */
+    private function handleGiftOption(Cart $cart, UpdateCartDeliverySettingsCommand $command): void
+    {
+        if ($command->isAGift() === null) {
+            return;
+        }
+
+        $cart->gift = $command->isAGift();
+
+        try {
+            if (false === $cart->update()) {
```

Comment:
There is a small issue, you can perform up to three calls to `$cart->update()` if three options are provided
The update should probably be handled in the `handle` method after those sub calls, either by checking if one of the three options is not null or by returning a boolean in these functions when an update is required

Reply:
Fixed in https://github.com/PrestaShop/PrestaShop/pull/21331/commits/2d6d2a7b4cfda0023ea2cef41026d1bf9c55ed68